src/org/sosy_lab/java_smt/test/SolverConcurrencyTest.java

```diff
@@ -0,0 +1,437 @@
+import static org.junit.Assert.assertTrue;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.junit.After;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+import org.sosy_lab.common.ShutdownManager;
+import org.sosy_lab.common.ShutdownNotifier;
+import org.sosy_lab.common.configuration.Configuration;
+import org.sosy_lab.common.configuration.InvalidConfigurationException;
+import org.sosy_lab.common.log.LogManager;
+import org.sosy_lab.common.rationals.Rational;
+import org.sosy_lab.java_smt.SolverContextFactory;
+import org.sosy_lab.java_smt.SolverContextFactory.Solvers;
+import org.sosy_lab.java_smt.api.BasicProverEnvironment;
+import org.sosy_lab.java_smt.api.BitvectorFormulaManager;
+import org.sosy_lab.java_smt.api.BooleanFormula;
+import org.sosy_lab.java_smt.api.BooleanFormulaManager;
+import org.sosy_lab.java_smt.api.FormulaManager;
+import org.sosy_lab.java_smt.api.IntegerFormulaManager;
+import org.sosy_lab.java_smt.api.Model;
+import org.sosy_lab.java_smt.api.NumeralFormula.IntegerFormula;
+import org.sosy_lab.java_smt.api.OptimizationProverEnvironment;
+import org.sosy_lab.java_smt.api.OptimizationProverEnvironment.OptStatus;
+import org.sosy_lab.java_smt.api.SolverContext;
+import org.sosy_lab.java_smt.api.SolverContext.ProverOptions;
+import org.sosy_lab.java_smt.api.SolverException;
+
+@RunWith(Parameterized.class)
+public class SolverConcurrencyTest {
+
+  private static final int NUMBER_OF_THREADS = 4;
+  private static final int TIMEOUT_SECONDS = 30;
+
+  private static List<Throwable> exceptionsList;
+
+  @Parameters(name = "{0}")
+  public static Object[] getAllSolvers() {
+    return Solvers.values();
+  }
+
+  @Parameter(0)
+  public Solvers solver;
+
+
+  protected Solvers solverToUse() {
+    return solver;
+  }
+
+  @BeforeClass
+  public static void createExceptionsList() {
+    exceptionsList = Collections.synchronizedList(new ArrayList<Throwable>());
+  }
+
+  @After
+  public void resetExceptionsList() {
+    exceptionsList.clear();
+  }
+
+  @Test
+  public void testIntConcurrencyWithConcurrentContext() {
+    // Test concurrency of integers (while every thread creates its unique context on its own
+    // concurrently)
+    assume().withMessage("Solver %s does not support the theory of bitvectors", solverToUse())
+        .that(solverToUse())
+        .isNotEqualTo(Solvers.BOOLECTOR);
+    List<Runnable> runnableList = new ArrayList<>();
+    for (int i = 0; i < NUMBER_OF_THREADS; i++) {
```

Comment:
Could you extract the loop into a utility method `List<Runable> getRunnables(Runnable r)`?
It might be possible to use Java lambdas for the instanciation of a runnable, e.g., something like
<pre>
Runnable test = () -> {
try {
￼            SolverContext context = initSolver();
￼            intConcurrencyTest(context);
￼            closeSolver(context);
￼          } catch (Throwable e) {
￼            exceptionsList.add(e);
￼          }
};
</pre>
This might improve readability.

Reply:
The first idea is problematic because we have different methods in the runnables every time.

I'll take a look into Lambdas but as far as i know that should be possible.

We can split the `List<Runnable> runnableList = new ArrayList<>();` into a `@BeforeClass` and reset after every test to increase readability further.